### PR TITLE
[Clang] Only ignore special methods for unused private fields in BuildFieldReferenceExpr

### LINKED ISSUE #116961: [Clang] False positive unused private field if defaulted member comparison operator declared out-of-class
Observed:
```c++
class A {
  int used;
  bool operator==(const A&) const;
};

bool A::operator==(const A&) const = default;

class B {
  int used;
  friend bool operator==(const B&, const B&);
};

bool operator==(const B&, const B&) = default;
```

`clang++ -std=c++20 -Wunused-private-field test.cpp`

[Compiler explorer link](https://godbolt.org/z/GnE3fdc45)

```plaintext
<source>:2:7: warning: private field 'used' is not used [-Wunused-private-field]
    2 |   int used;
      |       ^
1 warning generated.
Compiler returned: 0
```

Exptected:
No warnings.

### PR DESCRIPTION
**Pull request by Maetveis**
The original code assumed that only special methods might be defined as defaulted. Since C++20 comparison operators might be defaulted too, and we *do* want to consider those as using the fields of the class.

Fixes: #116961

### CODE REVIEW

File: clang/test/SemaCXX/warn-unused-private-field.cpp [hunk 1]
```diff
@@ -20,6 +20,20 @@ class SpaceShipDefaultCompare {
   int operator<=>(const SpaceShipDefaultCompare &) const = default;
 };
 
+class EqDefaultCompareOutOfClass {
+  int used;
```
Comment (erichkeane) [inline clang/test/SemaCXX/warn-unused-private-field.cpp:24]: A comment here pointing out that we're validating that `used` doesn't get diagnosed would be helpful.

**Review by erichkeane — COMMENTED** [review]
Needs a release note, else LGTM.

File: clang/test/SemaCXX/warn-unused-private-field.cpp [hunk 2]
```diff
@@ -20,6 +20,20 @@ class SpaceShipDefaultCompare {
   int operator<=>(const SpaceShipDefaultCompare &) const = default;
 };
 
+class EqDefaultCompareOutOfClass {
+  int used; // no warning
```
Comment (erichkeane) [inline clang/test/SemaCXX/warn-unused-private-field.cpp:24]: Please explain a touch better with 'no warning'?  Something like, "Test to ensure that the defaulted operator results in this being seen as 'used' despite there not being code to do so."?

Comment (Maetveis) [reply clang/test/SemaCXX/warn-unused-private-field.cpp]: Okay, sorry I wasn't sure what you were referring to, but I get know that from the change-set it is not obvious how / why the issue in #116961 is fixed by this change and why it is a special case not covered by the current testing. I expanded quite a bit on the comment, hopefully it did not become too verbose now.

**Review by erichkeane — APPROVED** [review]

**Pull request merged**